From 4f25b85b8a7798609c2c74badad82cc56c5adff3 Mon Sep 17 00:00:00 2001
From: Tommy Lin <lin.tommysc@inventec.com>
Date: Thu, 16 Feb 2023 05:29:30 +0000
Subject: [PATCH 21/27] Simple Rack-mounted Server for Sensors

Symptom/Reason:
- N/A

Root Cause:
- Simple Rack-mounted Server for Sensors

Solution/Change:
- [bmcweb]
  - Add PhysicalContext in Sensors
- [os-release]
  - Add BOOT_VERSION

Entry Test:
- Get sensor info
  - curl -u root:0penBmc -X GET https://${bmc}/redfish/v1/Chassis/${chassis_id}/Sensors/${sensor_id}
- Get bootloader version
  - curl -u root:0penBmc -X GET https://${bmc}/redfish/v1/Managers/bmc

---
 redfish-core/lib/sensors.hpp | 39 ++++++++++++++++++++++++++++++++++++
 1 file changed, 39 insertions(+)

diff --git a/redfish-core/lib/sensors.hpp b/redfish-core/lib/sensors.hpp
index ff764aad..c6e83ee2 100644
--- a/redfish-core/lib/sensors.hpp
+++ b/redfish-core/lib/sensors.hpp
@@ -187,6 +187,34 @@ inline std::string_view toReadingUnits(std::string_view sensorType)
     }
     return "";
 }
+
+inline const char* toPhysicalContext(const std::string& sensorName)
+{
+    std::string sensorNameLower =
+        boost::algorithm::to_lower_copy(sensorName);
+
+    if (sensorNameLower.find("dimm") != std::string::npos)
+    {
+        return "Memory";
+    }
+    if (sensorNameLower.find("psu") != std::string::npos)
+    {
+        return "PowerSupply";
+    }
+    if (sensorNameLower.find("fan") != std::string::npos)
+    {
+        return "Fan";
+    }
+    if (sensorNameLower.find("brd") != std::string::npos)
+    {
+        return "Board";
+    }
+    if (sensorNameLower.find("cpu") != std::string::npos)
+    {
+        return "CPU";
+    }
+    return "Chassis";
+}
 } // namespace sensors
 
 /**
@@ -800,6 +828,17 @@ inline void objectPropertiesToJson(
         {
             sensorJson["ReadingUnits"] = readingUnits;
         }
+
+        const std::string& physicalContext = sensors::toPhysicalContext(std::string(sensorName));
+        if (physicalContext.empty())
+        {
+            BMCWEB_LOG_ERROR << "Redfish cannot map physical context for "
+                             << sensorName;
+        }
+        else
+        {
+            sensorJson["PhysicalContext"] = physicalContext;
+        }
     }
     else if (sensorType == "temperature")
     {
-- 
2.34.1

